### ISSUE TITLE
Issue with downloading documents when file dir is a mounted share

### ISSUE THREAD

**Issue by bosquig** [glpi-project/glpi#12799]
### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

9.5.9

### Bug description

Hello,

We run glpi in a docker container, php version : 7.3.31 and the file dir (/var/www/html/glpi/files) is a symbolic link to a mounted volume which is an azure fileshare.

The issue : we can't display images attached to a ticket as document (or download other documents), except when the file is really tiny (like 2ko) : document.send.php returns a 400 without any logging. the last function called is readfile($file) which seems to be OK because it returns the right number of bytes read.
If we remove the link to the mounted share and use local files dir, we have no issue, images are displayed correctly !
As stated above, if the image is around 2ko, the image is displayed correctly even with the fileshare mounted.

Also, with the same infrastructure we have no issue at all on GLPI 9.3.1, so i looked for a difference between these 2 glpi versions and found what causes the issue in the code : 
inc/toolbox.class.php L883: 

`   // Make sure there is nothing in the output buffer (In case stuff was added by core or misbehaving plugin).
   // If there is any extra data, the sent file will be corrupted.
   while (ob_get_level() > 0) {
      ob_end_clean();
   }`

So if I remove this part, it all works fine. If I leave this part but that I add ob_start(); just before the readfile call, it also works.

tldr; when using a mounted fileshare for the file dir, displaying image document only works if output buffering is ON ; and with glpi 9.5.9 the output buffering is turned OFF with this call ob_end_clean(); L886

Do you have any idea why I got this issue in this specifc case and what I can do besides modifying glpi core ?
And if there is no other way, is it possible to turn on output buffering before the readfile call in a new glpi 9.5 release ?

Thanks in advance for your reply.
Grégory.

### Relevant log output

```shell
No output whatsoever in the error logs
```

### Your GLPI setup information

GLPI 9.5.9 ( => /var/www/html/glpi)
Installation mode: TARBALL
Current language:fr_FR

Server
 
Operating system: Linux a4ec993ccbdc 5.4.0-1089-azure #94~18.04.1-Ubuntu SMP Fri Aug 5 12:34:50 UTC 2022 x86_64
PHP 7.3.31-1~deb10u1 apache2handler (Core, PDO, Phar, Reflection, SPL, SimpleXML, Zend OPcache, apache2handler, apc, apcu, bz2,
	calendar, ctype, curl, date, dom, exif, fileinfo, filter, ftp, gd, gettext, hash, iconv, imap, intl, json, ldap, libxml,
	mbstring, mysqli, mysqlnd, openssl, pcre, pdo_mysql, posix, readline, session, shmop, sockets, sodium, standard, sysvmsg,
	sysvsem, sysvshm, tokenizer, wddx, xml, xmlreader, xmlrpc, xmlwriter, xsl, zip, zlib)
Setup: max_execution_time="30" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
	upload_max_filesize="2M" 
Software: Apache/2.4.38 (Debian) (Apache/2.4.38 (Debian) Server at web-qa-pst-glpi.azurewebsites.net Port 80)
	Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36 Edg/105.0.1343.50
Server Software: MariaDB Server
	Server Version: 10.3.23-MariaDB-log
	Server SQL Mode: 
	Parameters: itsm@maria-qa-pst-glpi@maria-qa-pst-glpi.mariadb.database.azure.com/itsm
	Host info: maria-qa-pst-glpi.mariadb.database.azure.com via TCP/IP
	
PHP version is at least 7.2.0 - Perfect!PHP version is at least 7.2.0 - Perfect!
Sessions support is available - Perfect!Sessions support is available - Perfect!
Allocated memory > 64 Mio - Perfect!Allocated memory > 64 Mio - Perfect!
mysqli extension is installedmysqli extension is installed
ctype extension is installedctype extension is installed
fileinfo extension is installedfileinfo extension is installed
json extension is installedjson extension is installed
mbstring extension is installedmbstring extension is installed
iconv extension is installediconv extension is installed
zlib extension is installedzlib extension is installed
curl extension is installedcurl extension is installed
gd extension is installedgd extension is installed
simplexml extension is installedsimplexml extension is installed
intl extension is installedintl extension is installed
ldap extension is installedldap extension is installed
apcu extension is installedapcu extension is installed
Zend OPcache extension is installedZend OPcache extension is installed
xmlrpc extension is installedxmlrpc extension is installed
exif extension is installedexif extension is installed
zip extension is installedzip extension is installed
bz2 extension is installedbz2 extension is installed
sodium extension is installedsodium extension is installed
Database version seems correct (10.3.23) - Perfect!Database version seems correct (10.3.23) - Perfect!
Timezones seems not loaded, see https://glpi-install.readthedocs.io/en/latest/timezones.html.Timezones seems not loaded, see https://glpi-install.readthedocs.io/en/latest/timezones.html.
The log file has been created successfully.The log file has been created successfully.
Write access to /var/www/html/glpi/files/_cache has been validated.Write access to /var/www/html/glpi/files/_cache has been validated.
Write access to /var/www/html/glpi/config has been validated.Write access to /var/www/html/glpi/config has been validated.
Write access to /var/www/html/glpi/files/_cron has been validated.Write access to /var/www/html/glpi/files/_cron has been validated.
Write access to /var/www/html/glpi/files has been validated.Write access to /var/www/html/glpi/files has been validated.
Write access to /var/www/html/glpi/files/_dumps has been validated.Write access to /var/www/html/glpi/files/_dumps has been validated.
Write access to /var/www/html/glpi/files/_graphs has been validated.Write access to /var/www/html/glpi/files/_graphs has been validated.
Write access to /var/www/html/glpi/files/_lock has been validated.Write access to /var/www/html/glpi/files/_lock has been validated.
Write access to /var/www/html/glpi/files/_pictures has been validated.Write access to /var/www/html/glpi/files/_pictures has been validated.
Write access to /var/www/html/glpi/files/_plugins has been validated.Write access to /var/www/html/glpi/files/_plugins has been validated.
Write access to /var/www/html/glpi/files/_rss has been validated.Write access to /var/www/html/glpi/files/_rss has been validated.
Write access to /var/www/html/glpi/sessions has been validated.Write access to /var/www/html/glpi/sessions has been validated.
Write access to /var/www/html/glpi/files/_tmp has been validated.Write access to /var/www/html/glpi/files/_tmp has been validated.
Write access to /var/www/html/glpi/files/_uploads has been validated.Write access to /var/www/html/glpi/files/_uploads has been validated.
Write access to /var/www/html/glpi/marketplace has been validated.Write access to /var/www/html/glpi/marketplace has been validated.
Web access to files directory is protectedWeb access to files directory is protected

GLPI constants
 
GLPI_ROOT: /var/www/html/glpi
GLPI_CONFIG_DIR: /var/www/html/glpi/config
GLPI_VAR_DIR: /var/www/html/glpi/files
GLPI_MARKETPLACE_DIR: /var/www/html/glpi/marketplace
GLPI_USE_CSRF_CHECK: 1
GLPI_CSRF_EXPIRES: 7200
GLPI_CSRF_MAX_TOKENS: 100
GLPI_USE_IDOR_CHECK: 1
GLPI_IDOR_EXPIRES: 7200
GLPI_ALLOW_IFRAME_IN_RICH_TEXT: 
GLPI_SERVERSIDE_URL_ALLOWLIST: Array
GLPI_TELEMETRY_URI: https://telemetry.glpi-project.org
GLPI_INSTALL_MODE: TARBALL
GLPI_NETWORK_MAIL: glpi@teclib.com
GLPI_NETWORK_SERVICES: https://services.glpi-network.com
GLPI_MARKETPLACE_PRERELEASES: 
GLPI_MARKETPLACE_ALLOW_OVERRIDE: 1
GLPI_MARKETPLACE_MANUAL_DOWNLOADS: 1
GLPI_USER_AGENT_EXTRA_COMMENTS: 
GLPI_AJAX_DASHBOARD: 1
GLPI_CALDAV_IMPORT_STATE: 0
GLPI_DEMO_MODE: 0
GLPI_FORCE_EMPTY_SQL_MODE: 1
GLPI_DOC_DIR: /var/www/html/glpi/files
GLPI_CACHE_DIR: /var/www/html/glpi/files/_cache
GLPI_CRON_DIR: /var/www/html/glpi/files/_cron
GLPI_DUMP_DIR: /var/www/html/glpi/files/_dumps
GLPI_GRAPH_DIR: /var/www/html/glpi/files/_graphs
GLPI_LOCAL_I18N_DIR: /var/www/html/glpi/files/_locales
GLPI_LOCK_DIR: /var/www/html/glpi/files/_lock
GLPI_LOG_DIR: /var/www/html/glpi/files/_log
GLPI_PICTURE_DIR: /var/www/html/glpi/files/_pictures
GLPI_PLUGIN_DOC_DIR: /var/www/html/glpi/files/_plugins
GLPI_RSS_DIR: /var/www/html/glpi/files/_rss
GLPI_SESSION_DIR: /var/www/html/glpi/sessions
GLPI_TMP_DIR: /var/www/html/glpi/files/_tmp
GLPI_UPLOAD_DIR: /var/www/html/glpi/files/_uploads
GLPI_NETWORK_REGISTRATION_API_URL: https://services.glpi-network.com/api/registration/
GLPI_MARKETPLACE_PLUGINS_API_URI: https://services.glpi-network.com/api/glpi-plugins/
GLPI_I18N_DIR: /var/www/html/glpi/locales
GLPI_VERSION: 9.5.9
GLPI_SCHEMA_VERSION: 9.5.7
GLPI_MIN_PHP: 7.2.0
GLPI_YEAR: 2022

Libraries
 
htmlawed/htmlawed version 1.2.5 in (/var/www/html/glpi/vendor/htmlawed/htmlawed)
phpmailer/phpmailer version 6.1.6 in (/var/www/html/glpi/vendor/phpmailer/phpmailer/src)
simplepie/simplepie version 1.5.6 in (/var/www/html/glpi/vendor/simplepie/simplepie/library)
tecnickcom/tcpdf version 6.3.5 in (/var/www/html/glpi/vendor/tecnickcom/tcpdf)
michelf/php-markdown in (/var/www/html/glpi/vendor/michelf/php-markdown/Michelf)
true/punycode in (/var/www/html/glpi/vendor/true/punycode/src)
iamcal/lib_autolink in (/var/www/html/glpi/vendor/iamcal/lib_autolink)
sabre/dav in (/var/www/html/glpi/vendor/sabre/dav/lib/DAV)
sabre/http in (/var/www/html/glpi/vendor/sabre/http/lib)
sabre/uri in (/var/www/html/glpi/vendor/sabre/uri/lib)
sabre/vobject in (/var/www/html/glpi/vendor/sabre/vobject/lib)
laminas/laminas-cache in (/var/www/html/glpi/vendor/laminas/laminas-cache/src)
laminas/laminas-i18n in (/var/www/html/glpi/vendor/laminas/laminas-i18n/src)
laminas/laminas-serializer in (/var/www/html/glpi/vendor/laminas/laminas-serializer/src)
monolog/monolog in (/var/www/html/glpi/vendor/monolog/monolog/src/Monolog)
sebastian/diff in (/var/www/html/glpi/vendor/sebastian/diff/src)
elvanto/litemoji in (/var/www/html/glpi/vendor/elvanto/litemoji/src)
symfony/console in (/var/www/html/glpi/vendor/symfony/console)
scssphp/scssphp in (/var/www/html/glpi/vendor/scssphp/scssphp/src)
laminas/laminas-mail in (/var/www/html/glpi/vendor/laminas/laminas-mail/src/Protocol)
laminas/laminas-mime in (/var/www/html/glpi/vendor/laminas/laminas-mime/src)
rlanvin/php-rrule in (/var/www/html/glpi/vendor/rlanvin/php-rrule/src)
blueimp/jquery-file-upload in (/var/www/html/glpi/vendor/blueimp/jquery-file-upload/server/php)
ramsey/uuid in (/var/www/html/glpi/vendor/ramsey/uuid/src)
psr/log in (/var/www/html/glpi/vendor/psr/log/Psr/Log)
psr/simple-cache in (/var/www/html/glpi/vendor/psr/simple-cache/src)
mexitek/phpcolors in (/var/www/html/glpi/vendor/mexitek/phpcolors/src/Mexitek/PHPColors)
guzzlehttp/guzzle in (/var/www/html/glpi/vendor/guzzlehttp/guzzle/src)
guzzlehttp/psr7 in (/var/www/html/glpi/vendor/guzzlehttp/psr7/src)
wapmorgan/unified-archive in (/var/www/html/glpi/vendor/wapmorgan/unified-archive/src)
paragonie/sodium_compat in (/var/www/html/glpi/vendor/paragonie/sodium_compat/src)
phpCas version 1.3.6 in (/usr/share/php/CAS/source)

**Comment by cedric-anne** [comment]
Hi,

Can you try following patch ?

```diff
diff --git a/src/Toolbox.php b/src/Toolbox.php
index cac9734e10..5eec2ff119 100644
--- a/src/Toolbox.php
+++ b/src/Toolbox.php
@@ -686,12 +686,17 @@ class Toolbox
         $etag = md5_file($file);
         $lastModified = filemtime($file);
 
-       // Make sure there is nothing in the output buffer (In case stuff was added by core or misbehaving plugin).
-       // If there is any extra data, the sent file will be corrupted.
-        while (ob_get_level() > 0) {
+        // Make sure there is nothing in the output buffer (In case stuff was added by core or misbehaving plugin).
+        // If there is any extra data, the sent file will be corrupted.
+        // 1. Turn off any extra buffuring level. Keep one buffering level if PHP output_buffering directive is "on".
+        $max_buffering_level = (int)ini_get('output_buffering');
+        while (ob_get_level() > $max_buffering_level) {
             ob_end_clean();
         }
-       // Now send the file with header() magic
+        // 2. Clean any buffered output in remaining level (output_buffering="on" case).
+        ob_clean();
+
+        // Now send the file with header() magic
         header("Last-Modified: " . gmdate("D, d M Y H:i:s", $lastModified) . " GMT");
         header("Etag: $etag");
         header_remove('Pragma');
```

**Comment by bosquig** [comment]
> 

Hello @cedric-anne and thanks for your reply.

Your patch works fine, but note that output_buffering is not necessarily a boolean and can be an integer that represents the maximum number of bytes (e.g 4096). So maybe it would be better to have something like that to ob_end_clean til last level : 
$max_buffering_level = (int)ini_get('output_buffering') > 0 ? 1 : 0;

Also, what is the purpose of ob_end_clean if output_buffering is set to OFF anyway in php.ini ?

If we're good on the patch, can it get published in a 9.x release ?

**Comment by cedric-anne** [comment]
> Your patch works fine, but note that output_buffering is not necessarily a boolean and can be an integer that represents the maximum number of bytes (e.g 4096). So maybe it would be better to have something like that to ob_end_clean til last level :
> $max_buffering_level = (int)ini_get('output_buffering') > 0 ? 1 : 0;

Indeed, I added a commit to handle this.

> Also, what is the purpose of ob_end_clean if output_buffering is set to OFF anyway in php.ini ?

Purpose is to handle the fact that a buggy plugin may have open one or more levels of buffers using `ob_start();`. As we send a file, we want to remove all that can be present in any OB level, to ensure that it will not be sent in the response.

> If we're good on the patch, can it get published in a 9.x release ?

9.x is not supported anymore. We only published latest releases to fix critical security issues.
